README.md
# NAME

WebService::OANDA::ExchangeRates - A Perl wrapper for the OANDA Exchange Rates
API

# SYNOPSIS

    my $api = WebService::OANDA::ExchangeRates->new(api_key => 'YOUR_API_KEY');

    # all API methods return a response object

    # get_currencies
    my $response = $api->get_currencies();
    if ($response->is_success) {
        print $response->data->{USD}; # US Dollar
    }
    else {
      print $response->error_message # an error message
    }

    # get the number of quotes remaining
    $response = $api->get_remaining_quotes();
    print $response->data->{remaining_quotes} if $response->is_success;

    # get a series of quotes for a base currency
    $response = $api->get_rates(
        base_currency => 'USD',
        quote         => [ qw{ EUR CAD } ],
    );
    if ($response->is_success) {
        my $base_currency = $quote->data->{base_currency};
        print "Base Currency: $base_currency\n";
        foreach my $quote_currency (keys %{$response->data->{quotes}) {
            my $quote = $response->data->{quotes}{$quote_currency};
            print "  $quote_currency:\n";
            print "    BID: ", $quote->{bid}, "\n";
            print "    ASK: ", $quote->{ask}, "\n";
        }
    }

# DESCRIPTION

This module provides a simple wrapper around the
[OANDA Exchange Rates API](http://www.oanda.com/rates) using [LWP::UserAgent](https://metacpan.org/pod/LWP::UserAgent).
Go to the
[API documentation page](http://developer.oanda.com/exchange-rates-api) for a
full reference of all the methods. This service requires you to
[sign up](http://www.oanda.com/rates/#pricing) for a trial or paying
subscription to obtain an API key.

# METHODS

## Constructor

- $api = WebService::OANDA::ExchangeRates->new(%options)

    The constructor for the API wrapper.  Other than `api_key` all other parameters
    are optional.

    - api\_key

        **REQUIRED** - the api key provided by the service

    - base\_url

            base_url => 'https://www.oanda.com/rates/api/v1/'

        The base url of the service. By default this is set to
        [https://www.oanda.com/rates/api/v1/](https://www.oanda.com/rates/api/v1/).  If your environment requires the
        service to be at a static IP address you can use
        [https://web-services.oanda.com/rates/api/v1/](https://web-services.oanda.com/rates/api/v1/).

    - proxy

            proxy => 'http://your.proxy.com:8080/'

        If you access the service behind a proxy, you can provide it. This sets the
        proxy for the underlying [LWP::UserAgent](https://metacpan.org/pod/LWP::UserAgent) object. Similarly, if the
        `PERL_LWP_ENV_PROXY` environment variable is set with the proxy, it will be
        used automatically.

    - timeout

            timeout => 180

        Set the maximum time for a request to return on the underlying [LWP::UserAgent](https://metacpan.org/pod/LWP::UserAgent)
        object. The default is 180 seconds.

## API Methods

All API methods return a [WebService::OANDA::ExchangeRates::Response](https://metacpan.org/pod/WebService::OANDA::ExchangeRates::Response) object
that provides access to the [HTTP::Response](https://metacpan.org/pod/HTTP::Response) object and has deserialized the
JSON response into a native Perl data structure.

### get\_currencies

- $response = $api->get\_currencies()

    Returns the `/v1/currencies.json` endpoint; a hash of valid currency codes.

    **NOTE:** This endpoint usually returns an array of hashes which contain _code_
    and _description_ keys but `get_currencies()` massages them into a hash with
    the _code_ as the key and _description_ as the value.

    The data structure returned by `$response->data` will look something like this:

        {
            USD => 'US Dollar',
            EUR => 'Euro',
            CAD => 'Canadian Dollar',
            ...
        }

### get\_remaining\_quotes

- $response = $api->get\_remaining\_quotes()

    Returns the `/v1/remaining_quotes.json` endpoint; the number of quote requests
    available in the current billing period.

    Some plans are limited to a specific number of quotes per billing period.  This
    endpoint can be used to determine how many quotes you have left.

    The data structure returned by `$response->data` will look something like this:

        {
          remaining_quotes => 100000,
        }

    For plans that have no quote limits, _remaining\_quotes_ will equal "unlimited".

### get\_rates

- $response = $api->get\_rates(%options)

    Returns the `/v1/rates/XXX.json` endpoint; a list of quotes for a specific base
    currency.

    This is the meat of the API and provides daily averages, highs, lows and
    midpoints for a each cross of currencies.  Only one argument is required,
    _base\_currency_.

    Please see the [OANDA Exchange Rates API Docs](http://developer.oanda.com/exchange-rates-api)
    for a breakdown of the returned data as it is a rather large data structure.

    - base\_currency

        **REQUIRED** - The base currency that all quotes are crossed against. Must be a
        valid 3 letter upper case currency code as provided by `/v1/currencies.json`
        endpoint.  This wrapper will not validate the currency code, just that it is
        in the correct format.

            base_currency => 'USD'

    - quote

        A single currency code, or an arrayref of currency codes to cross against the
        _base\_currency_.

        **DEFAULT:** All other available currencies.

            quote => 'EUR'
            quote => [qw{ EUR GBP CHF }]

    - decimal\_places

        The number of decimal places to provide in the quote. May be a positive integer
        of reasonable size (as of this writing, up to 14) or the string "all".  Quotes
        that are requested with more precision than exist are padded out with zero's.

        **DEFAULT:** 5

            decimal_places => 'all'

    - fields

        Which fields to return in the quotes. This can be specified as a single string
        or an arrayref of strings.

        As of this writing, fields can be any of the following:

        - averages - the bid and ask
        - midpoint - the midpoint between the bid and ask
        - highs - the highest bid and ask
        - lows - the lowest bid and ask

        It should be noted that this module does not restrict what these strings are
        so as to be forward compatible with any changes.

        **DEFAULT:** averages

            fields => 'all'
            fields => [qw{ averages midpoint }],

    - date

        The requested date for the quotes. This must be in _YYYY-MM-DD_ format.  The
        24 hour period of the date is considered to be that period in UTC.

        **DEFAULT:** The most recent quote

            date => '2014-02-01'

    - start, end

        This allows you to specify a date range. Also in _YYYY-MM-DD_ format.  When
        requesting a date range, quotes are modified such that:

        - averages (bid and ask) are the average of the daily values over the date range
        - midpoint (midpoint) is the midpoint between those averages
        - highs (high\_ask and high\_bid) are the highest values over the range
        - lows (low\_ask and low\_bid) are the lowest values over the range

        Specifying no `end` will assume today's date as the end point. Date ranges are
        inclusive (they include all quotes on and between `start` and `end`).

        **DEFAULT:** none

            start => '2014-01-01',
            end   => '2014-01-31',

## Accessors

### api\_key

> The originally supplied api\_key.

### base\_url

> Returns a [URI](https://metacpan.org/pod/URI) object

### proxy

> Returns a [URI](https://metacpan.org/pod/URI) object

### timeout

> The request timeout

### user\_agent

- $api->user\_agent

    This provides access to the underlying [LWP::UserAgent](https://metacpan.org/pod/LWP::UserAgent) instance.

    **NOTE:** While you can certainly modify the object to accomodate needed special
    settings, removing the default headers will cause authentication to fail.

# SEE ALSO

- [HTTP::Response](https://metacpan.org/pod/HTTP::Response)
- [LWP::UserAgent](https://metacpan.org/pod/LWP::UserAgent)
- [URI](https://metacpan.org/pod/URI)
- [WebService::OANDA::ExchangeRates::Response](https://metacpan.org/pod/WebService::OANDA::ExchangeRates::Response)
- [OANDA Exchange Rates API](http://www.oanda.com/rates) landing page
- [OANDA Exchange Rates API Docs](http://developer.oanda.com/exchange-rates-api)

# SUPPORT

## Bug/Feature requests

Please file a ticket at the repository for this code on github at:

[https://github.com/oanda/perl-webservice-oanda-exchangerates](https://github.com/oanda/perl-webservice-oanda-exchangerates)

# AUTHOR

    Dave Doyle <ddoyle@oanda.com>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by OANDA Corporation.

This program is free software; you can redistribute it and/or modify it under the same terms as Perl itself.


Changes
Revision history for {{$dist->name}}

{{$NEXT}}

0.003     2014-07-16 11:27:11-04:00 America/Toronto
        - new parameter 'data_set' for use with get_rates and get_currencies
        - provides access to new European Central Bank rates

0.002     2014-05-08 09:23:52-04:00 America/Toronto
        - Hopefully fix up failing tests on cpantesters

0.001     2014-05-07 15:38:45-04:00 America/Toronto
        - Initial release


lib/WebService/OANDA/ExchangeRates/Response.pm
package WebService::OANDA::ExchangeRates::Response;

use Moo;

use HTTP::Response;
use JSON::XS qw{decode_json};
use Try::Tiny;

has http_response => (
    is  => 'rw',
    handles => {
        is_success  => 'is_success',
        is_error    => 'is_error',
        raw_data    => 'content',
        http_status => 'code',
    },
    required => 1,
);

has data => ( is  => 'rw' );

sub BUILD {
    my $self = shift;
    my $data;

    try {
        $data = JSON::XS::decode_json( $self->raw_data );
    }
    catch {
        $data = {
            code => undef,
            message => sprintf(
                'Failed to decode content(%s): $%s', $self->raw_data, $_
            )
        };

        # if we failed to decode the json on a successful query, we'll consider
        # this an outlier and force a 500
        if ($self->is_success) {
            $self->http_response(
                HTTP::Response->new(500, undef, undef, $self->raw_data)
            );
        }

    };

    $self->data($data);
}

sub error_message {
    my $self = shift;

    return unless $self->is_error;
    return $self->data->{message};
}

sub error_code {
    my $self = shift;
    return unless $self->is_error;
    return $self->data->{code};
}

1;

__END__

=head1 NAME

WebService::OANDA::ExchangeRates::Response - the response object from all API
calls in L<WebService::OANDA::ExchangeRates>.

=head1 SYNOPSIS

  my $api = WebService::OANDA::ExchangeRates->new(api_key => 'YOUR_API_KEY');

  my $response = $api->get_quotes(
      base_currency => 'USD',
      quote         => [ qw{ EUR CAD } ],
  );

  if ($response->is_success) {
      print $response->data->{base_currency}, "\n";
  }
  else {
      print $respone->error_message, "\n";
  }



=head1 DESCRIPTION

This module is used as the result generated by any API method called in
L<WebService::OANDA::ExchangeRates>.  It automatically deserializes the data
into a native Perl data structure and provides information about the request
and response via an instance of L<HTTP::Response>.  You should never need to
instantiate this directly as it will be provided.

=head1 METHODS

=head2 Constructor

=over 4

=item $response = WebService::OANDA::ExchangeRates::Response->new( http_response => $http_response_object );

Constructor takes just an L<HTTP::Response> object which will attempt to
deserialize the JSON into a native data structure.

=back

=head2 data

Contains the deserialized data structure of the API request.

=head2 error_code

A convenience method. If the response was an error, returns the contents of
C<< $response->data->{code} >>.

=head2 error_message

A convenience method. If the response was an error, returns the contents of
C<< $response->data->{message} >>.

=head2 http_response

Accessor for the L<HTTP::Response> object generated by the API request.

=head2 http_status

Convenience method to access C<< $response->http_response->code >> to determine
the HTTP status code.

=head2 is_error

=head2 is_success

Convenience methods that proxy through to the same methods of the
C<< $response->http_response >> instance.

=head2 raw_data

Convenience method to examine the raw serialized data returned by the API.  A
proxied request to C<< $response->http_response->content >>.

=head1 SEE ALSO

=over 4

=item * L<HTTP::Response>

=item * L<WebService::OANDA::ExchangeRates>

=back

=head1 SUPPORT

=head2 Bug/Feature requests

Please file a ticket at the repository for this code on github at:

L<https://github.com/oanda/perl-webservice-oanda-exchangerates>

=head1 AUTHOR

  Dave Doyle <ddoyle@oanda.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by OANDA Corporation.

This program is free software; you can redistribute it and/or modify it under the same terms as Perl itself.


lib/WebService/OANDA/ExchangeRates.pm
package WebService::OANDA::ExchangeRates;

use JSON::XS;
use LWP::UserAgent;
use Moo;
use Types::Standard qw{ ArrayRef Int Str StrMatch };
use Type::Utils qw{ declare as where coerce from via enum};
use Types::URI qw{ Uri };
use WebService::OANDA::ExchangeRates::Response;

use vars qw($VERSION);
$VERSION = "0.002";

has base_url => (
    is       => 'ro',
    isa      => Uri,
    coerce   => Uri->coercion,
    default  => 'https://www.oanda.com/rates/api/v1/',
    required => 1,
);
has proxy => (
    is        => 'ro',
    isa       => Uri,
    coerce    => Uri->coercion,
    predicate => 1,
);
has timeout => ( is => 'ro', isa => Int, predicate => 1 );
has api_key => ( is => 'ro', isa => Str, required  => 1 );
has user_agent => ( is => 'ro', lazy => 1, builder => 1, init_arg => undef );

has _rates_validator => ( is => 'ro', required => 1, builder => 1 );

sub _build__rates_validator {

    # TYPES FOR VALIDATOR
    my $Currency = declare as StrMatch[qr{^[A-Z]{3}$}];
    my $Quotes   = declare as ArrayRef[$Currency];
    coerce $Quotes, from Str, via { [$_] };
    my $Date = declare as StrMatch[qr{^\d\d\d\d-\d\d-\d\d$}];
    my $Fields = declare as ArrayRef[Str];
    coerce $Fields, from Str, via { [$_] };
    my $DecimalPlaces = StrMatch[qr{^(?:\d+|all)$}];

    my %validators = (
        base_currency   => $Currency,
        quote           => $Quotes,
        date            => $Date,
        start           => $Date,
        end             => $Date,
        fields          => $Fields,
        decimal_places  => $DecimalPlaces,
    );

    return sub {
        my %params = @_;

        my %result = ();
        die "missing required parameter: base_currency"
            unless exists $params{base_currency};

        foreach my $key ( keys %params ) {
            my $type = $validators{$key};
            die "invalid parameter: $key" unless $type;
            my $val = $params{$key};
            $val = $type->coerce($val) if $type->has_coercion;
            if ( ! $type->check($val) ) {
                $val = JSON::XS::encode_json($val) if ref $val;
                die "invalid value: $key = ($val)";
            }
            $result{$key} = $val;
        }
        return \%result;

    }
}

sub _build_user_agent {
    my $self = shift;

    my %options = ( agent => sprintf '%s/%s', __PACKAGE__, $VERSION );

    # LWP::UA forces hostname verification by default in newer versions
    # turn off for simplification;
    $options{ssl_opts} = { verify_hostname => 0 }
        if $self->base_url->scheme eq 'https';
    $options{timeout} = $self->timeout if $self->has_timeout;

    my $ua = LWP::UserAgent->new(%options);

    # set auth header
    $ua->default_header(
        Authorization => sprintf 'Bearer %s', $self->api_key
    );

    # set the proxy if needed
    # LWP:UserAgent will use PERL_LWP_ENV_PROXY if set automatically
    $ua->proxy( $self->base_url->scheme, $self->proxy ) if $self->has_proxy;

    return $ua;
}

# GET /currencies.json
sub get_currencies {
    my $self = shift;

    my $response = $self->_get_request('currencies.json');

    # convert arrayref[hashref] into hashref
    if ( $response->is_success && exists $response->data->{currencies}) {
        $response->data({
            map { $_->{code} => $_->{description} }
                @{$response->data->{currencies}}
        });
    }

    return $response;
}

# GET /rates/XXX.json
sub get_rates {
    my $self = shift;
    my $params = $self->_rates_validator->(@_);

    my $base_currency = delete $params->{base_currency};
    return $self->_get_request(['rates', $base_currency . '.json'], $params);
}

# GET /remaining_quotes.json
sub get_remaining_quotes {
    my $self = shift;
    return $self->_get_request('remaining_quotes.json');
}

sub _get_request {
    my ( $self, $path, $params ) = @_;

    my $uri = $self->base_url->clone->canonical;

    # build the new path
    $path = [$path] unless ref $path eq 'ARRAY';
    my @new_path = grep { $_ ne '' } ($uri->path_segments, @{$path});
    $uri->path_segments( @new_path );

    # set query params
    $params = {} unless defined $params;
    $uri->query_form(%{$params});

    my $response = $self->user_agent->get( $uri->as_string );
    return WebService::OANDA::ExchangeRates::Response->new(
        http_response => $response );
}

1;

__END__

=head1 NAME

WebService::OANDA::ExchangeRates - A Perl wrapper for the OANDA Exchange Rates
API

=head1 SYNOPSIS

  my $api = WebService::OANDA::ExchangeRates->new(api_key => 'YOUR_API_KEY');

  # all API methods return a response object

  # get_currencies
  my $response = $api->get_currencies();
  if ($response->is_success) {
      print $response->data->{USD}; # US Dollar
  }
  else {
    print $response->error_message # an error message
  }

  # get the number of quotes remaining
  $response = $api->get_remaining_quotes();
  print $response->data->{remaining_quotes} if $response->is_success;

  # get a series of quotes for a base currency
  $response = $api->get_rates(
      base_currency => 'USD',
      quote         => [ qw{ EUR CAD } ],
  );
  if ($response->is_success) {
      my $base_currency = $quote->data->{base_currency};
      print "Base Currency: $base_currency\n";
      foreach my $quote_currency (keys %{$response->data->{quotes}) {
          my $quote = $response->data->{quotes}{$quote_currency};
          print "  $quote_currency:\n";
          print "    BID: ", $quote->{bid}, "\n";
          print "    ASK: ", $quote->{ask}, "\n";
      }
  }

=head1 DESCRIPTION

This module provides a simple wrapper around the
L<OANDA Exchange Rates API|http://www.oanda.com/rates> using L<LWP::UserAgent>.
Go to the
L<API documentation page|http://developer.oanda.com/exchange-rates-api> for a
full reference of all the methods. This service requires you to
L<sign up|http://www.oanda.com/rates/#pricing> for a trial or paying
subscription to obtain an API key.

=head1 METHODS

=head2 Constructor

=over 4

=item $api = WebService::OANDA::ExchangeRates->new(%options)

The constructor for the API wrapper.  Other than C<api_key> all other parameters
are optional.

=over 4

=item * api_key

B<REQUIRED> - the api key provided by the service

=item * base_url

  base_url => 'https://www.oanda.com/rates/api/v1/'

The base url of the service. By default this is set to
L<https://www.oanda.com/rates/api/v1/>.  If your environment requires the
service to be at a static IP address you can use
L<https://web-services.oanda.com/rates/api/v1/>.

=item * proxy

  proxy => 'http://your.proxy.com:8080/'

If you access the service behind a proxy, you can provide it. This sets the
proxy for the underlying L<LWP::UserAgent> object. Similarly, if the
C<PERL_LWP_ENV_PROXY> environment variable is set with the proxy, it will be
used automatically.

=item * timeout

  timeout => 180

Set the maximum time for a request to return on the underlying L<LWP::UserAgent>
object. The default is 180 seconds.

=back

=back

=head2 API Methods

All API methods return a L<WebService::OANDA::ExchangeRates::Response> object
that provides access to the L<HTTP::Response> object and has deserialized the
JSON response into a native Perl data structure.

=head3 get_currencies

=over 4

=item $response = $api->get_currencies()

Returns the C</v1/currencies.json> endpoint; a hash of valid currency codes.

B<NOTE:> This endpoint usually returns an array of hashes which contain I<code>
and I<description> keys but C<get_currencies()> massages them into a hash with
the I<code> as the key and I<description> as the value.

The data structure returned by C<< $response->data >> will look something like this:

  {
      USD => 'US Dollar',
      EUR => 'Euro',
      CAD => 'Canadian Dollar',
      ...
  }

=back

=head3 get_remaining_quotes

=over 4

=item $response = $api->get_remaining_quotes()

Returns the C</v1/remaining_quotes.json> endpoint; the number of quote requests
available in the current billing period.

Some plans are limited to a specific number of quotes per billing period.  This
endpoint can be used to determine how many quotes you have left.

The data structure returned by C<< $response->data >> will look something like this:

  {
    remaining_quotes => 100000,
  }

For plans that have no quote limits, I<remaining_quotes> will equal "unlimited".

=back

=head3 get_rates

=over 4

=item $response = $api->get_rates(%options)

Returns the C</v1/rates/XXX.json> endpoint; a list of quotes for a specific base
currency.

This is the meat of the API and provides daily averages, highs, lows and
midpoints for a each cross of currencies.  Only one argument is required,
I<base_currency>.

Please see the L<OANDA Exchange Rates API Docs|http://developer.oanda.com/exchange-rates-api>
for a breakdown of the returned data as it is a rather large data structure.

=over 4

=item * base_currency

B<REQUIRED> - The base currency that all quotes are crossed against. Must be a
valid 3 letter upper case currency code as provided by C</v1/currencies.json>
endpoint.  This wrapper will not validate the currency code, just that it is
in the correct format.

  base_currency => 'USD'

=item * quote

A single currency code, or an arrayref of currency codes to cross against the
I<base_currency>.

B<DEFAULT:> All other available currencies.

  quote => 'EUR'
  quote => [qw{ EUR GBP CHF }]

=item * decimal_places

The number of decimal places to provide in the quote. May be a positive integer
of reasonable size (as of this writing, up to 14) or the string "all".  Quotes
that are requested with more precision than exist are padded out with zero's.

B<DEFAULT:> 5

  decimal_places => 'all'

=item * fields

Which fields to return in the quotes. This can be specified as a single string
or an arrayref of strings.

As of this writing, fields can be any of the following:

=over 4

=item * averages - the bid and ask

=item * midpoint - the midpoint between the bid and ask

=item * highs - the highest bid and ask

=item * lows - the lowest bid and ask

=back

It should be noted that this module does not restrict what these strings are
so as to be forward compatible with any changes.

B<DEFAULT:> averages

  fields => 'all'
  fields => [qw{ averages midpoint }],

=item * date

The requested date for the quotes. This must be in I<YYYY-MM-DD> format.  The
24 hour period of the date is considered to be that period in UTC.

B<DEFAULT:> The most recent quote

  date => '2014-02-01'

=item * start, end

This allows you to specify a date range. Also in I<YYYY-MM-DD> format.  When
requesting a date range, quotes are modified such that:

=over 4

=item * averages (bid and ask) are the average of the daily values over the date range

=item * midpoint (midpoint) is the midpoint between those averages

=item * highs (high_ask and high_bid) are the highest values over the range

=item * lows (low_ask and low_bid) are the lowest values over the range

=back

Specifying no C<end> will assume today's date as the end point. Date ranges are
inclusive (they include all quotes on and between C<start> and C<end>).

B<DEFAULT:> none

  start => '2014-01-01',
  end   => '2014-01-31',

=back

=back

=head2 Accessors

=head3 api_key

=over 4

The originally supplied api_key.

=back

=head3 base_url

=over 4

Returns a L<URI> object

=back

=head3 proxy

=over 4

Returns a L<URI> object

=back

=head3 timeout

=over 4

The request timeout

=back

=head3 user_agent

=over 4

=item $api->user_agent

This provides access to the underlying L<LWP::UserAgent> instance.

B<NOTE:> While you can certainly modify the object to accomodate needed special
settings, removing the default headers will cause authentication to fail.

=back

=head1 SEE ALSO

=over 4

=item * L<HTTP::Response>

=item * L<LWP::UserAgent>

=item * L<URI>

=item * L<WebService::OANDA::ExchangeRates::Response>

=item * L<OANDA Exchange Rates API|http://www.oanda.com/rates> landing page

=item * L<OANDA Exchange Rates API Docs|http://developer.oanda.com/exchange-rates-api>

=back

=head1 SUPPORT

=head2 Bug/Feature requests

Please file a ticket at the repository for this code on github at:

L<https://github.com/oanda/perl-webservice-oanda-exchangerates>

=head1 AUTHOR

  Dave Doyle <ddoyle@oanda.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by OANDA Corporation.

This program is free software; you can redistribute it and/or modify it under the same terms as Perl itself.
